### ISSUE TITLE
Q: Optional TraversalType Declaration

### ISSUE THREAD

**Issue by ts-smith** [facebook/relay#145]
I was thinking about a potential client optimization for relay, and was wondering how feasible it would be. I am currently thinking that abusing the node interface may get some of the way there. I am still a bit hazy about how relay works so I may be way off base here.

Imagine a universe of types {A, B, C, Root}.

with some shorthand type definitions:

```
A {
   C: {
      resolve: (obj) => resolveC(obj.c_id),
   }
}
B {
    C:{
       resolve: (obj) => resolveC(obj.c_id)
    }
}

Root {
   ...,
   C: {
       resolve: (obj, args) => resolveC(args.id)
   }
}
```

I believe that relay is able to reason about the traversals to C from any of the individual originating nodes, but not comprehensively between them. The type contract is only applied on the nodes, but what if it could be applied to traversals (optionally) as well?

something like

```
var optionOverride = {
   traversalType: 'identityTraversal',
   identifyingField: '_id',
}

var traversalType = createTraversalType(CType, optionOverride);

@shorthand
A  {
    C: {
          traversalType,
          resolve: (obj) => resolveC(obj.c_id)
    }
}
...
```

which, when compiled into the schema, could help relay identify objects at the end of edges better. I was thinking of using the node interface in my fragments to do that part of the way, but that seems like it would be really messy in practice.

**Comment by josephsavona** [comment]
Can you provide a specific example (data) and what type of data access you want to optimize? It's not clear what the goal is.

Note that Relay can directly refetch any object that implements `Node`, via its `id` and the `node_id: $id)` root field defined in the spec.

**Comment by ts-smith** [comment]
I realized that the thing I proposed would not be able to work in a significant number of cases, so it probably wouldn't be worth the trouble. But the originating problem comes from having multiple root queries that don't seem to be able to share the cached results, because the arguments aren't standardized (I think). For example, I may have a query like

```
query(selectedB: $selectedB} {
    A {
        Bs {
            ..someFields
        }
    },
    B(id: $selectedB) @include(if: $selectedB) {
        ...manyFields
    }
}
```

I think that since the actual B root query has not been seen before if it is "selected" for the first time, the query builder won't know that it has already seen the node. I was thinking I could do a node query for B

```
    ....
    node(id: $selectedB) @include(if: $selectedB) {
        ...on B {
            ...manyFields
        }
    }
```

instead, so that the standardized node arguments would facilitate optimized query generation.

**Comment by josephsavona** [comment]
This can occur at any level of the hierarchy and is not specific to roots; in general given some new query, it's possible that the client happens to have fetched a subset of that data. For example, given the query `node(id: 1) { foo { bar { ... } } }`, the client may have already fetched the `bar` record via `node(id: bar) {...}` or via `node(id: foo) { bar { ... } }`, or from some other completely unrelated node that happens to link to `bar`. However, there is no way to avoid refetching this subset of data without the server knowing *exactly* what has been fetched by the client.

In short, it turns out to be impractical to try and avoid this category of refetching. Instead, Relay will detect if the results of a new query overlap with existing records. When this happens the new results are merged into the cache, and only components whose data actually changed will re-render. 

Note that typical navigation patterns - moving from a "list of things" screen to an item detail screen - means that in practice this type of refetching is quite rare.

**Comment by ts-smith** [comment]
I see. My suggestion was then, I suppose, to flatten the queries as much as possible on the client so that each layer can be checked locally. Then you would have to descend through the graph layer by layer, which might make it so less fetching is done in total, but would induce a communication overhead with added complexity which would probably defeat the purpose.

**Comment by josephsavona** [comment]
Yeah, ultimately it's better to trade a moderate amount of refetching to achieve the performance win of a single round trip. 

I'm gong to close this out, but thanks for the great discussion and keep up the suggestions!
